### DISCUSSION TITLE
TextBlock Inlines Binding

### DISCUSSION THREAD

**Discussion by H0tCh0colat3** [AvaloniaUI/Avalonia#7972] [Q&A] [answered]
What technically prevents us from being able to bind to a text block's Inlines? I have a use case requiring that a text block with many different styles and font sizes display text one character at a time to simulate text being typed on a screen. The solution I came up with involves updating the Text property one character at a time with a timer delay. It works well enough for a plain text block, but not for one with different styles.

Once I saw that the Inlines was unbindable, I tried breaking the text down into individually styled characters (one per textblock) and binding the character collection to a wrap panel, which sort of works, except I lose proper word wrapping. This could be slightly better if I instead broke the text into styled chunks based on whitespace or punctuation, but even then, different font sizes aren't aligned to the same base line if they are in different text blocks.

If I've missed some vastly better way of doing this, please point me in the right direction.

Just curious if binding to Inlines is at all feasible, or if it would take a monumental amount of work to do so.

Edit: For clarity, this is a hard-coded example of what I want.
![with hardcoded inlines](https://user-images.githubusercontent.com/32641956/163082118-f9c93cf0-7767-4422-b870-876e35705347.PNG)

and this was the result of my attempt to break the text into individual characters. This is not what I want
![with hardcoded wrap panel](https://user-images.githubusercontent.com/32641956/163082112-aa8c6636-6cd2-47e6-a566-dc0752449832.PNG)

**Comment by Gillibald** [comment, marked as answer]
The InlineCollection needs to be created by the TextBlock and therefor it is not possible to make it bindable but you can use a behavior to manipulate the collection via binding

https://stackoverflow.com/questions/1959856/data-binding-the-textblock-inlines

**Comment by AlexeyTliss** [comment]
Necrobump!
Not quire sure why Avalonia team didn't implement it, so here is a snippet for all the people who want to bind Inlines:
```
public static class TextBlock
{
	static TextBlock()
	{
		InlinesProperty.Changed.Subscribe(InlinesSourceChanged);
	}

	#region Inlines

	private static readonly Dictionary<Avalonia.Controls.TextBlock, IDisposable> _inlinesSources = [];

	public static readonly AttachedProperty<INotifyCollectionChanged> InlinesProperty =
		AvaloniaProperty.RegisterAttached<Avalonia.Controls.TextBlock, INotifyCollectionChanged>("Inlines", typeof(TextBlock));

	public static INotifyCollectionChanged GetInlines(Avalonia.Controls.TextBlock textBlock) => textBlock.GetValue(InlinesProperty);
	public static void SetInlines(Avalonia.Controls.TextBlock textBlock, INotifyCollectionChanged value) => textBlock.SetValue(InlinesProperty, value);

	private static void InlinesSourceChanged(AvaloniaPropertyChangedEventArgs<INotifyCollectionChanged> args)
	{
		if (args.Sender is not Avalonia.Controls.TextBlock textBlock)
			return;

		if (args.NewValue.HasValue && args.NewValue.Value != null)
			SubscribeInlinesSource(textBlock, args.NewValue.Value);
		else
			UnsubscribeInlinesSource(textBlock, null);
	}

	private static void SubscribeInlinesSource(Avalonia.Controls.TextBlock textBlock, INotifyCollectionChanged inlinesSource)
	{
		UnsubscribeInlinesSource(textBlock, null);
		_inlinesSources[textBlock] = inlinesSource.ObserveCollectionChanges().Subscribe(_ => UpdateInlines(textBlock, inlinesSource));
		textBlock.DetachedFromLogicalTree += UnsubscribeInlinesSource;
		Dispatcher.UIThread.Post(() => UpdateInlines(textBlock, inlinesSource));
	}

	private static void UnsubscribeInlinesSource(object sender, LogicalTreeAttachmentEventArgs _)
	{
		var textBlock = (Avalonia.Controls.TextBlock)sender;
		textBlock.DetachedFromLogicalTree -= UnsubscribeInlinesSource;
		if (_inlinesSources.Remove(textBlock, out var inlinesObservable))
			inlinesObservable.Dispose();
	}

	private static void UpdateInlines(Avalonia.Controls.TextBlock textBlock, INotifyCollectionChanged inlinesSource)
	{
		var sources = ((System.Collections.ICollection)inlinesSource).Cast<object>().ToList();
		var addCount = sources.Count - textBlock.Inlines.Count;
		var removeCount = textBlock.Inlines.Count - sources.Count;

		for (int i = 0; addCount > 0 && i < addCount; i++)
			textBlock.Inlines.Add(new TemplatedControl());

		if (removeCount > 0)
			textBlock.Inlines.RemoveRange(textBlock.Inlines.Count - removeCount, removeCount);

		for (int i = 0; i < sources.Count; i++)
		{
			textBlock.Inlines[i].DataContext = null;  // по какой-то причине Авалония не обновляет биндинги стилей (классов) корректно, если не сбросить DataContext на null принудительно
			textBlock.Inlines[i].DataContext = sources[i];
		}
	}

	#endregion
}
```

How to use:
- Put TextBlock.cs above in MyProject/Behaviors/Textblock.cs
- add xmlns:behaviors="using:Dictation.Common.UI.Behaviors" in your UserControl or Window
- in TextBlock or SelectableTextBlock use: behaviors:TextBlock.Inlines="{Binding MyViewModelItems}"
- MyViewModelItems must be ObservableCollection<T> or it's descendant in your viewmodel (or at least it should implement INotifyCollectionChanged)
XAML sample:
```
<SelectableTextBlock  behaviors:TextBlock.Inlines="{Binding Texts}"
				TextWrapping="Wrap">
				<SelectableTextBlock.Styles>
					<Style Selector="TemplatedControl">
						<Setter Property="Template">
							<ControlTemplate>
								<Border Margin="0 3"
										ToolTip.Tip="{Binding ToolTip}"
										ZIndex="0">
									<TextBlock Margin="2 0"
											   VerticalAlignment="Center"
											   Text="{Binding Word}" />
								</Border>
							</ControlTemplate>
						</Setter>
					</Style>
				</SelectableTextBlock.Styles>
</SelectableTextBlock>
```

If you want selection to work on each symbol, change `new TemplatedControl()` to `new Run()`, but you won't be able to use dynamic template then, only text decorations

Cheers!
